robots/flakefinder/downloader.go

```diff
@@ -0,0 +1,202 @@
+func listGcsObjects(ctx context.Context, client *storage.Client, bucketName, prefix, delim string) (
+	[]string, error) {
+
+	var objects []string
+	it := client.Bucket(bucketName).Objects(ctx, &storage.Query{
+		Prefix:    prefix,
+		Delimiter: delim,
+	})
+
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return objects, fmt.Errorf("error iterating: %v", err)
+		}
+
+		if attrs.Prefix != "" {
+			objects = append(objects, path.Base(attrs.Prefix))
+		}
+	}
+	logrus.Info("end of listGcsObjects(...)")
+	return objects, nil
+}
+
+func readGcsObject(ctx context.Context, client *storage.Client, bucket, object string) ([]byte, error) {
+	logrus.Infof("Trying to read gcs object '%s' in bucket '%s'\n", object, bucket)
+	o := client.Bucket(bucket).Object(object)
+	reader, err := o.NewReader(ctx)
+	if err == storage.ErrObjectNotExist {
+		return nil, err
+	} else if err != nil {
+		return nil, fmt.Errorf("cannot read object '%s': %v", object, err)
+	}
+	return ioutil.ReadAll(reader)
+}
+
+func FindUnitTestFiles(ctx context.Context, client *storage.Client, bucket, repo string, pr *github.PullRequest) ([]*Result, error) {
+
+	dirOfPrJobs := path.Join("pr-logs", "pull", strings.ReplaceAll(repo, "/", "_"), strconv.Itoa(pr.Number))
+
+	prJobsDirs, err := listGcsObjects(ctx, client, bucket, dirOfPrJobs+"/", "/")
+	if err != nil {
+		return nil, fmt.Errorf("error listing gcs objects: %v", err)
+	}
+
+	junits := []*Result{}
+	for _, job := range prJobsDirs {
+		junit, err := FindUnitTestFileForJob(ctx, client, bucket, dirOfPrJobs, job, pr)
+		if err != nil {
+			return nil, err
+		}
+		if junit != nil {
+			junits = append(junits, junit...)
+		}
+	}
+	return junits, err
+}
+
+func FindUnitTestFileForJob(ctx context.Context, client *storage.Client, bucket string, dirOfPrJobs string, job string, pr *github.PullRequest) ([]*Result, error) {
+	dirOfJobs := path.Join(dirOfPrJobs, job)
+
+	prJobs, err := listGcsObjects(ctx, client, bucket, dirOfJobs+"/", "/")
+	if err != nil {
+		return nil, fmt.Errorf("error listing gcs objects: %v", err)
+	}
+	builds := sortBuilds(prJobs)
+	profilePath := ""
+	buildNumber := 0
+	reports := []*Result{}
+	for _, build := range builds {
+		buildDirPath := path.Join(dirOfJobs, strconv.Itoa(build))
+		dirOfFinishedJSON := path.Join(buildDirPath, finishedJSON)
+		dirOfStartedJSON := path.Join(buildDirPath, startedJSON)
+
+		_, err := readGcsObject(ctx, client, bucket, dirOfFinishedJSON)
+		if err == storage.ErrObjectNotExist {
+			// build still running?
+			continue
+		} else if err != nil {
+			return nil, fmt.Errorf("Cannot read finished.json (%s) in bucket '%s'", dirOfFinishedJSON, bucket)
+		} else {
+			startedJSON, err := readGcsObject(ctx, client, bucket, dirOfStartedJSON)
+			if err != nil {
+				return nil, fmt.Errorf("Cannot read started.json (%s) in bucket '%s'", dirOfStartedJSON, bucket)
+			}
+			if !IsLatestCommit(startedJSON, pr) {
+				break
```

Comment:
continue?

Reply:
No, as I saw this is correct as is. I'm currently debugging some kind of mapping problem there.